### ISSUE TITLE
How to consume the result object of Either in Use Case Layer?

### ISSUE THREAD

**Issue by crjacinro** [android10/Android-CleanArchitecture-Kotlin#72]
So there is a repository layer and a user case layer. My use case calls the repository method (which returns either left or right and returns the User object as output when right. 
`
    override suspend fun run(params: Params): Either<Failure, User> {
        val result = userRepository.loginUser(params.username, params.password)
        
        // How to get the result object "User" in this line here? So that I can save it into the database (for example)
        
        return result.right(User())
    }
`

**Comment by crjacinro** [comment]
Hi. Any ideas?

**Comment by crjacinro** [comment]
Not sure if this is correct but I can get it through result.either call and provide left or right callbacks, in which in the right callback, I can receive the User param.

**Comment by Zhuinden** [comment]
Have you tried `.fold()`?

**Comment by crjacinro** [comment]
> Have you tried `.fold()`?

Nope, that method is not available in either.

**Comment by ferdyrod** [comment]
You could use `.flatMap()`. it gives access to the Right object.

`result.flatMap() { it -> 
    repository.saveUser(it) }
`

**Comment by Zhuinden** [comment]
Hmm according to Arrow's Either implementation, there should be a `fold` method.

``` kotlin
  inline fun <R> fold(ifLeft: (LEFT) -> R, ifRight: (RIGHT) -> R): R = when (this) {
    is Right -> ifRight(right)
    is Left -> ifLeft(left)
  }
```

**Comment by ferdyrod** [comment]
> Hmm according to Arrow's Either implementation, there should be a `fold` method.
> 

Yes, Arrow's implementation does have `fold` method. But this project has a custom Either, it doesn't have all of Arrow methods. It only has `flatmap()` and `map().`

**Comment by Zhuinden** [comment]
Well it's kinda neutered if it cannot fold. That's kinda the whole point of either 😄

**Comment by crjacinro** [comment]
> You could use `.flatMap()`. it gives access to the Right object.
> 
> `result.flatMap() { it -> repository.saveUser(it) } `

Thanks for this but doesn't flatmap transform data (instead of accessing object only) ? what is the impact of the flatmap operation to the original result variable?

**Comment by crjacinro** [comment]
Also upon checking the code, what is the difference between map and flat map?

**Comment by Zhuinden** [comment]
`flatMap` expects some kind of `Either<L, R>` while `map` in case of `Either` most likely means `mapRight` which expects some kind of `R`.
